File: java/leetCode/binarySearch/Problem034.java
package leetCode.binarySearch;

/**
 * Given an array of integers nums sorted in ascending order,
 * find the starting and ending position of a given target value.
 * Your algorithm's runtime complexity must be in the order of O(log n).
 * If the target is not found in the array, return [-1, -1].
 * <p>
 * Example 1:
 * Input: nums = [5,7,7,8,8,10], target = 8
 * Output: [3,4]
 * <p>
 * Example 2:
 * Input: nums = [5,7,7,8,8,10], target = 6
 * Output: [-1,-1]
 * <p>
 * create by stephen on 2018/5/5
 */
public class Problem034 {

    /**
     * 找到target第一次和最后一次出现的位置即可
     */
    public static int[] searchRange(int[] nums, int target) {
        if (nums.length == 0) return new int[]{-1, -1};
        int left = 0, right = nums.length - 1, first;

        //找到第一次出现的位置
        while (left < right) {
            int mid = left + ((right - left) >> 1);
            if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        if (nums[left] != target) {
            return new int[]{-1, -1};
        }
        first = left;

        left = 0;
        right = nums.length - 1;

        //找到最后一次出现的位置
        while (left < right) {
            int mid = right - ((right - left) >> 1);
            if (nums[mid] <= target) {
                left = mid;
            } else {
                right = mid - 1;
            }
        }

        return new int[]{first, left};
    }
}

File: leetCode/binSearch/Problem034.cpp
#include <functional>
#include <vector>

/**
 *
 * Given an array of integers nums sorted in ascending order, find the starting
 * and ending position of a given target value.
 *
 * If target is not found in the array, return [-1, -1].
 *
 * Follow up: Could you write an algorithm with O(log n) runtime complexity?
 */
int bisect(std::vector<int>& nums, std::function<bool(int)>& f) {
  int s = 0, e = nums.size();
  while (s < e) {  // first element >= target
    int m = (s + e) >> 1;
    if (!f(nums[m]))
      s = m + 1;
    else
      e = m;
  }
  return s;
}

std::vector<int> searchRange(std::vector<int>& nums, int target) {
  if (nums.size() == 0) return {-1, -1};
  int s = bisect(nums, [target](int a) { return a >= target; });
  if (s == nums.size() || nums[s] != target) return {-1, -1};
  int e = bisect(nums, [target](int a) { return a > target; }) - 1;
  return {s, e};
}